cmd/kubectl-gadget/top/block-io.go

```diff
@@ -121,164 +175,96 @@ func blockIOCallback(line string, node string) {
 		return
 	}
 
-	blockIONodeStats[node] = event.Stats
+	p.nodeStats[node] = event.Stats
 }
 
-func blockIOStartPrintLoop() {
+func (p *BlockIOParser) StartPrintLoop() {
 	go func() {
-		ticker := time.NewTicker(time.Duration(outputInterval) * time.Second)
-		blockIOPrintHeader()
+		ticker := time.NewTicker(time.Duration(p.flags.OutputInterval) * time.Second)
+		p.PrintHeader()
 		for {
 			_ = <-ticker.C
-			blockIOPrintHeader()
-			blockIOPrintEvents()
+			p.PrintHeader()
+			p.PrintStats()
 		}
 	}()
 }
 
-func blockIOPrintHeader() {
-	switch params.OutputMode {
-	case commonutils.OutputModeColumns:
-		if term.IsTerminal(int(os.Stdout.Fd())) {
-			utils.ClearScreen()
-		} else {
-			fmt.Println("")
-		}
+func (p *BlockIOParser) PrintHeader() {
+	if p.OutputConfig.OutputMode == commonutils.OutputModeJSON {
+		return
+	}
 
-		fmt.Printf("%-16s %-16s %-16s %-16s %-7s %-16s %-3s %-6s %-6s %-7s %-8s %s\n",
-			"NODE", "NAMESPACE", "POD", "CONTAINER",
-			"PID", "COMM", "R/W", "MAJOR", "MINOR", "BYTES", "TIME(µs)", "IOs")
-	case commonutils.OutputModeCustomColumns:
-		if term.IsTerminal(int(os.Stdout.Fd())) {
-			utils.ClearScreen()
-		} else {
-			fmt.Println("")
-		}
-		fmt.Println(blockIOGetCustomColsHeader(params.CustomColumns))
+	if term.IsTerminal(int(os.Stdout.Fd())) {
+		utils.ClearScreen()
+	} else {
+		fmt.Println("")
 	}
+
+	fmt.Println(p.BuildColumnsHeader())
 }
 
-func blockIOPrintEvents() {
-	// sort and print events
-	mutex.Lock()
+func (p *BlockIOParser) PrintStats() {
+	// Sort and print stats
+	p.Lock()
 
 	stats := []types.Stats{}
-	for _, stat := range blockIONodeStats {
+	for _, stat := range p.nodeStats {
 		stats = append(stats, stat...)
 	}
-	blockIONodeStats = make(map[string][]types.Stats)
+	p.nodeStats = make(map[string][]types.Stats)
 
-	mutex.Unlock()
+	p.Unlock()
 
-	types.SortStats(stats, blockIOSortBy)
+	types.SortStats(stats, p.flags.ParsedSortBy)
 
-	switch params.OutputMode {
-	case commonutils.OutputModeColumns:
-		for idx, event := range stats {
-			if idx == maxRows {
-				break
-			}
-
-			rw := 'R'
-			if event.Write {
-				rw = 'W'
-			}
-
-			fmt.Printf("%-16s %-16s %-16s %-16s %-7d %-16s %-3c %-6d %-6d %-7d %-8d %d\n",
-				event.Node, event.Namespace, event.Pod, event.Container,
-				event.Pid, event.Comm, rw, event.Major, event.Minor, event.Bytes,
-				event.MicroSecs, event.Operations)
-		}
-	case commonutils.OutputModeJSON:
-		b, err := json.Marshal(stats)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "Error: %s", commonutils.WrapInErrMarshalOutput(err))
-			return
-		}
-		fmt.Println(string(b))
-	case commonutils.OutputModeCustomColumns:
-		for idx, stat := range stats {
-			if idx == maxRows {
-				break
-			}
-			fmt.Println(blockIOFormatEventCustomCols(&stat, params.CustomColumns))
+	for idx, stat := range stats {
+		if p.OutputConfig.OutputMode != commonutils.OutputModeJSON && idx == p.flags.MaxRows {
```

Comment:
I was wondering the same when I modified this code. We were already applying this feature only for non-JSON output mode.

I can enable it for the JSON output mode as well but I think we should improve the description of this flag because "Maximum rows to print" doesn't make sense for the JSON output mode. WDYT?

Reply:
I think the confusion is because "row" = "item" in the columns output mode. I don't have ideas about to make this description better, without renaming it to something else. I think it's just fine to reduce the number of generated items in the json if you're passing `--maxRows` and `-o json`.